### PR TITLE
V4

### PR DESCRIPTION
**Pull request by i-like-robots**
It's SVG all the way down.

The proposed implementation for icons is as follows:

```html
<i class="o-ft-icon o-ft-icon--{size} o-ft-icon--{icon} o-ft-icon--{colour}"></i>
```

It was decided that keeping a distinct HTML element was preferable to using pseudo-content to append icons to other elements. This is because it is more explicit and (hopefully) more end-user-friendly, as well as increasing the options for implementation.

sprites

This is the full sprite sheet consisting of all the currently available icons and basic o-colors colour palette. It weighs in at ~130kb uncompressed and ~40kb gzipped making it ~10kb larger than the equivelent icon font. The fallback .png is around ~230kb and ~210kb gzipped at the default resolution and 128 colours. The actual size of the .png will differ depending on usage.

## How does it work?

The `.o-ft-icon` base class will transform the element it is applied to into a 'viewport' and create pseudo-content to be used as a canvas. The the icon sprite is applied to the canvas as a background image. The canvas is scaled according to the number of icons and colours it contains and absolutely positioned relative to the viewport element.

implementation-detail

> "Why would you not just use a background image like we did in 2008 instead of this pseudo-content nonsense?!"

Chill. Because the background position X and Y coordinates cannot be manipulated independently. It [was proposed](http://www.w3.org/Style/CSS/Tracker/issues/9) but dropped from the spec. If I'd used `background-position` I'd need to generate combinator classes for each colour/icon pair - that's 1000 extra declaration blocks for 50 icons and 20 colours.

## What about the fallback?

Browsers that don't support this implementation fall into two categories:

### Browsers that don't support SVG but do support CSS3 background image options (Android 2.x)

The fallback is simple, we just provide the .png image and it can be scaled using `background-size`.

### Old IE

IE doesn't support background sizing. The ActiveX filter `AlphaImageLoader` does have a `sizingMethod` property but unfortunately the only options available are not suitable for spriting (equivalent to `background-size: 100% 100%;`). It also [does not apply to psuedo-content](http://stackoverflow.com/a/10669986).

I explored generating sprites for individual colours and icons but due to the scaling issues these approaches were not viable. 

To avoid generating hundreds, if not thousands of extra CSS declarations the IE fallback is tied to the `.o-ft-icon--{size}` class or `oFtIconsIconSize()` mixin which will load a scaled background .png image just for IE. This only creates extra declarations and requests for the number of sizes configured, which shouldn't be many.

**UPDATE**: using the `zoom` property https://github.com/Financial-Times/o-ft-icons/pull/101

## I don't want all of the icons/colours. Do I have to have them?

For now, yes. I have explored creating a service that provides only the requested icons/colors which [can be found here](https://github.com/i-like-robots/svg-sprite-generator).

sprite-service

This version of o-ft-icons must be configured using Sass variables in order to generate the correct classes for the available icons and colours. It is possible these things could be tied together in the future.

## Where are the non-square icons?

These are gone. All icons from now on will be on a square canvas. The non-square icons were limited to logos and branding which will be moved into a separate branding image set. These should be used as images and not for stylistic purposes.

## What does it look like then?

Chrome
chrome

Firefox
firefox

Safari
safari

Edge
edge

IE10
ie10

IE8
ie8

Tested on Android 2.x and iOS Safari 7 as well.

## TO-DO

- ~~Public mixins~~
- ~~Some documentation~~
- ~~Make sure multiple images aren't accidentally downloaded~~
- ~~Update fallback mixin to use image service so fallback sprites don't have to be provided~~

### CODE REVIEW

**Comment by wheresrhys** [comment]
TODO - renaming icons?

**Comment by AlbertoElias** [comment]
@wheresrhys Removing, renaming and adding new icons will be a separate task. This is just about how the module should work.

**Comment by wheresrhys** [comment]
but renaming icons in future would be a breaking change would it not? This PR is called V4 so IMO anything atht shodl go into the next major should go in here

**Comment by AlbertoElias** [comment]
@wheresrhys this is a merge into the v4 branch, there will be more work before the final release is made, including that, sorry for not explaining properly!

**Comment by wheresrhys** [comment]
fair enough

**Comment by i-like-robots** [comment]
> Excellent write up.

> Regarding this point:

> > I don't want all of the icons/colours. Do I have to have them?
>Maybe not a service but perhaps we could a have few separate SVGs that group related icons together? For example: social-icons.svg, section-icons.svg, arrow-icons.svg.

> These could be turned on and off individually in Sass. eg $o-ft-icons-social-icons: true

I haven't explored grouping icons as a possibility but it may be worth doing post icon clean up. At my last job we split icons into two groups; 'core' and... the rest. The rest had no fallback.

File: scss/_variables.scss [hunk 1]
```diff
@@ -7,17 +7,97 @@
...
+	'pink-tint3': #cec6b9,
+	'pink-tint4': #a7a59b,
+	'pink-tint5': #74736c
+) !default;
+
+/// Icon sizes
+/// @type Map
+$o-ft-icons-icon-sizes: (
+	'xs': 16,
+	's':	24,
```
Comment (AlbertoElias) [inline scss/_variables.scss]: Spacing

**Pull request merged**